### PR TITLE
Fix the setting of the proxy in the request module to ensure that it …

### PR DESCRIPTION
**Pull request by mmurphy**
Fix the setting of the proxy in the request module to ensure that it is set after the config is already loaded.

The way it was being set when each module was loaded was prone to errors, if the module was referenced before the config had been loaded.
This came to light due to the version module being loaded before the config and so the proxy was not configured in the version module
which is used before all commands, so all all commands were broken if using a proxy.

Jira: https://issues.jboss.org/browse/FH-2856

### CODE REVIEW

File: lib/cmd/common/version.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ const PATTERN = /\+BUILD-NUMBER/;
 var fhc = require('../../fhc'),
   ini = require('../../utils/ini'),
   _ = require('underscore'),
-  request = require('request').defaults({'proxy': fhc.config.get("proxy"), timeout: timeout}),
+  request = require('request'),   // don't set proxy or other configured defaults here, since the config may not be loaded when this module is
```
Comment (fh-qube) [inline lib/cmd/common/version.js]: INFO Move this trailing comment on the previous empty line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingComment)

File: lib/cmd/fh2/app/act.js [hunk 2]
```diff
@@ -10,7 +10,7 @@ var fhreq = require("../../../utils/request");
 var common = require("../../../common.js");
 var util = require('util');
 var ini = require('../../../utils/ini');
-var request = require('request').defaults({'proxy': fhc.config.get("proxy")});
+var request = require('request');     // don't set proxy or other configured defaults here, since the config may not be loaded when this module is
```
Comment (fh-qube) [inline lib/cmd/fh2/app/act.js]: INFO Move this trailing comment on the previous empty line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingComment)

File: lib/cmd/fhc/login.js [hunk 3]
```diff
@@ -11,8 +11,7 @@ var crypto;
 var targets = require("./targets.js");
 var fhreq = require("../../utils/request");
 var ngui = require("../common/ngui");
-var proxyConfig = require('../../utils/proxy');
-var request = proxyConfig.configureProxy(require('request'), fhc.config.get('proxy'));
+var request = require('request');   // don't set proxy or other configured defaults here, since the config may not be loaded when this module is
```
Comment (fh-qube) [inline lib/cmd/fhc/login.js]: INFO Move this trailing comment on the previous empty line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingComment)

File: lib/cmd/fh2/init.js [hunk 4]
```diff
@@ -28,8 +28,9 @@ module.exports = function init(argv, callback) {
   // We use stream type console output to prevent automatic newline.
   process.stdout.write(i18n._("Fetching project ZIP from GitHub... "));
 
+  var proxyRequest = request.defaults({'proxy': fhc.config.get("proxy")})
```
Comment (fh-qube) [inline lib/cmd/fh2/init.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/cmd/internal/df.js [hunk 5]
```diff
@@ -18,7 +18,7 @@ df.usage = "\nfhc df ping"
 var log = require("../../utils/log");
 var fhc = require("../../fhc");
 var util = require('util');
-var request = require('request').defaults({'proxy': fhc.config.get("proxy")});
+var request = require('request');   // don't set proxy or other configured defaults here, since the config may not be loaded when this module is
```
Comment (fh-qube) [inline lib/cmd/internal/df.js]: INFO Move this trailing comment on the previous empty line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingComment)

File: lib/cmd/internal/df.js [hunk 6]
```diff
@@ -108,7 +110,8 @@ function apps(cb) {
 }
 
 function logs(appName, cb) {
-  request(getDynoFarmUrl() + "logs/" + appName, function (err, response, body) {
+  var proxyRequest = request.defaults({'proxy': fhc.config.get("proxy")})
```
Comment (fh-qube) [inline lib/cmd/internal/df.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/cmd/internal/df.js [hunk 7]
```diff
@@ -190,7 +196,8 @@ function getDynoFarmUrl() {
 
 // delete an app
 function deleteApp(appName, cb) {
-  request({url: getDynoFarmUrl() + "delete/" + appName, method: 'POST'}, function (err, response, body) {
+  var proxyRequest = request.defaults({'proxy': fhc.config.get("proxy")})
```
Comment (fh-qube) [inline lib/cmd/internal/df.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/cmd/internal/df.js [hunk 8]
```diff
@@ -117,14 +120,16 @@ function logs(appName, cb) {
...
+  proxyRequest(getDynoFarmUrl() + "stop/" + appName, function (err, response, body) {
     return cb(common.nullToUndefined(err), body);
   });
 }
 
 // Start app
 function start(appName, cb) {
-  request(getDynoFarmUrl() + "start/" + appName, function (err, response, body) {
+  var proxyRequest = request.defaults({'proxy': fhc.config.get("proxy")})
```
Comment (fh-qube) [inline lib/cmd/internal/df.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/cmd/internal/df.js [hunk 9]
```diff
@@ -90,7 +91,8 @@ function createTableForApps(apps) {
 
 // DynoFarm apps
 function apps(cb) {
-  request(getDynoFarmUrl() + "apps", function (err, response, body) {
+  var proxyRequest = request.defaults({'proxy': fhc.config.get("proxy")})
```
Comment (fh-qube) [inline lib/cmd/internal/df.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/cmd/internal/df.js [hunk 10]
```diff
@@ -117,14 +120,16 @@ function logs(appName, cb) {
 }
 
 function stop(appName, cb) {
-  request(getDynoFarmUrl() + "stop/" + appName, function (err, response, body) {
+  var proxyRequest = request.defaults({'proxy': fhc.config.get("proxy")})
```
Comment (fh-qube) [inline lib/cmd/internal/df.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/cmd/fh3/app/_baseCloudRequest.js [hunk 11]
```diff
@@ -4,7 +4,7 @@ var read = require('../../fh2/app/read.js').read;
 var log = require("../../../utils/log");
 var util = require('util');
 var fhc = require("../../../fhc");
-var request = require('request').defaults({'proxy': fhc.config.get("proxy")});
+var request = require('request');     // don't set proxy or other configured defaults here, since the config may not be loaded when this module is
```
Comment (fh-qube) [inline lib/cmd/fh3/app/_baseCloudRequest.js]: INFO Move this trailing comment on the previous empty line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingComment)

**Comment by fh-qube** [comment]
SonarQube analysis reported 14 issues:
* BLOCKER 8 blocker
* INFO 6 info

Watch the comments in this conversation to review them.

**Comment by fh-automatron** [comment]
### Pull Request Quality Guidelines

As we continue to focus on quality, here are a few things to keep in mind:

* The code changes conform to the contributing & code style guides for this component (if they exist)
* A component owner should review your change (we've tagged these above)
* Automated testing has been added or updated for this change (e.g. UART, Unit/Acceptance tests)
* If updating server component this PR should be deployed to a grid somewhere (via `fhcap-cli`) - please provide a link to this
* Ensure JIRA is up to date before & after merge (e.g. deployment notes, tester notes, release notes, fix version)
* Any SonarQube issues should be addressed where applicable (use your own judgement)

Please ping a component owner (see `fh contrib`) when your builds are complete, changes are deployed and you're ready for a review.
Fhcap PR will be generated automatically.
To deploy your changes to a cluster place comment in this PR with pattern [x][y] where x=deploy and y=clustername you wish to deploy. 
You can also deploy your change manually using fhcap-cli

**Comment by wtrocki** [comment]
Reviewed changes. It looks like we changing this proxy everywhere we need to. 
I was interested if we can wrap this into some internal require but it would be tricky - we would need to call the method on usage rather than refer to object.

Do you have any proxy configured at the moment (can be sent as PM) 
I would like to verify this by setting proxy:
` fhc fhcfg set proxy youproxy.com`

**Comment by mmurphy** [comment]
Thanks @wtrocki , I do have a VM with a proxy.
However setting the proxy config is not sufficient to test, you would need to block access to the internet, except through the proxy, on your test machine, in order to ensure that all requests are being proxied.
I do also have a VM setup for that.

I can give you access if you want to test it - details also in the Jira.

**Comment by wtrocki** [comment]
Verified. Good to merge :)

**Pull request merged**